units/src/amount/error.rs

```diff
@@ -389,3 +389,40 @@ impl fmt::Display for PossiblyConfusingDenominationError {
 impl std::error::Error for PossiblyConfusingDenominationError {
     fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
 }
+
+/// An error consensus decoding an `Amount`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum DecodeError {
```

Comment:
I may be over-thinking all this, but I believe this is a good example of why I'm struggling with different strategies for the `Either` type. Here we have an error type with a general decoding "bytes level" variant (eof) and a specific type variant (out of range). It feels clean to keep these together vs. putting them in two layers (kinda like what is going on here with the legacy [ParseError](https://github.com/rust-bitcoin/rust-bitcoin/blame/558896bc1d73b4a32bddd94e2290efa68acedcf6/bitcoin/src/consensus/error.rs#L153) which looks to hold more of the lower level stuff). However, the combo'ing of all these error types in the composite decoders (Decoder2/3/4/6) creates quite the nest for a caller to deal with, can imagine these will get quite deep for a transaction.

Reply:
We should decide have some uniformity in the error names.

- `amount::DecoderError`
- `AmountDecoderError`
- `DecodeError` vs `DecoderError`

I'm leaning towards `amount::DecoderError` (to match the trait name).

Then for example in `transaction` we'd have:

- `TxOutDecoderError`, `TransactionDecoderError` etc

Reply:
And make a call on `non_exhaustive`.

Reply:
These should all be nonexhaustive. Always room for us to provide more helpful error information. Arguably they should be totally opaque types.

As for names, I have no opinion. Agreed we should be consistent. I like using the decoder name with `Error` at the end.

Reply:
@nyonson want to weigh in on the name? Otherwise I'll go for Andrew's vote e.g., `AmountDecoderError`, `OutPointDecoderError`.

Reply:
Sounds good to me!